docs/unit-testing.md

```diff
@@ -0,0 +1,211 @@
+# Unit Testing
+
+This section documents ways to facilitate unit testing when using Batcher. [Testify](https://github.com/stretchr/testify) was used in the examples below, but it is not a requirement.
+
+- [Using mocks](#using-mocks)
+  - [AzureSharedResource](#azuresharedresource)
+  - [ProvisionedResource](#provisionedresource)
+  - [RateLimiter](#ratelimiter)
+- [Using events](#using-events)
+
+## Using mocks
+
+There are public interfaces provided for Batcher, Watcher, Operation, AzureSharedResource, ProvisionedResource, and RateLimiter to use for mocking. To mock using Testify, you can follow this pattern:
+
+1. Implement the mock interface (for example, a mock Watcher):
+
+    ```go
+    import (
+        "testing"
+        "time"
+
+        gobatcher "github.com/plasne/go-batcher/v2"
+        "github.com/stretchr/testify/assert"
+        "github.com/stretchr/testify/mock"
+    )
+
+    type mockWatcher struct {
+        mock.Mock
+    }
+
+    func (w *mockWatcher) WithMaxAttempts(val uint32) gobatcher.Watcher {
+        args := w.Called(val)
+        return args.Get(0).(gobatcher.Watcher)
+    }
+
+    func (w *mockWatcher) WithMaxBatchSize(val uint32) gobatcher.Watcher {
+        args := w.Called(val)
+        return args.Get(0).(gobatcher.Watcher)
+    }
+
+    func (w *mockWatcher) WithMaxOperationTime(val time.Duration) gobatcher.Watcher {
+        args := w.Called(val)
+        return args.Get(0).(gobatcher.Watcher)
+    }
+
+    func (w *mockWatcher) MaxAttempts() uint32 {
+        args := w.Called()
+        return args.Get(0).(uint32)
+    }
+
+    func (w *mockWatcher) MaxBatchSize() uint32 {
+        args := w.Called()
+        return args.Get(0).(uint32)
+    }
+
+    func (w *mockWatcher) MaxOperationTime() time.Duration {
+        args := w.Called()
+        return args.Get(0).(time.Duration)
+    }
+
+    func (w *mockWatcher) ProcessBatch(batch []gobatcher.Operation) {
+        w.Called(batch)
+    }
+    ```
+
+1. Write a test method mocking any calls that are used in the underlying methods (for example, Enqueue calls MaxAttempts):
+
+    ```go
+    func TestEnqueueOnlyOnce(t *testing.T) {
+        batcher := gobatcher.NewBatcher()
+        watcher := &mockWatcher{}
+        watcher.On("MaxAttempts").Return(uint32(1))
+        op := gobatcher.NewOperation(watcher, 0, struct{}{}, false)
+        op.MakeAttempt() // fake a previous attempt
+        err := batcher.Enqueue(op)
+        assert.Error(t, err, "should only be allowed one time")
+    }
+    ```
+
+### AzureSharedResource
+
+One of the configuration options for AzureSharedResource is `WithMocks()`. For unit testing, you can pass mocks to AzureSharedResource to emulate an Azure Storage Account (specifically a mock blob and a mock container). This allows you to unit test without needing a real Azure Storage Account.
```

Comment:
I mentioned this in another PR, but I have reservations about this mechanism for injecting mocks. I've not seen it done this way in any other codebase, and it strikes me as odd that using a testing backdoor like this would be a preferred approach.

Why not just use more traditional dependency injection techniques? Constructor injection, Property/Member injection, Factory/Provider, even just a simple application of Strategy pattern all seem like they would work here?

I'm hoping to learn some new aspect of this approach that makes it appealing, perhaps it's more idiomatic for Go or those other approaches have downsides in Go?

Reply:
Additional Note on my previous comment:

I realize that this PR is about the `docs` and not about the `code`, but reading this description of how to consume the code really crystallized exactly what it was about the code that seemed off.

Reply:
Let's talk about those options, because I don't really see them as different than injecting the mocks using WithMocks().

Reply:
This discussion has been moved over to the other PR: https://github.com/plasne/go-batcher/pull/19#discussion_r614839221